baseapp/baseapp.go

```diff
@@ -615,83 +620,62 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 	// MultiStore in case message processing fails. At this point, the MultiStore
 	// is doubly cached-wrapped.
 	runMsgCtx, msCache := app.cacheTxContext(ctx, txBytes)
-	result = app.runMsgs(runMsgCtx, msgs, mode)
-	result.GasWanted = gasWanted
-
-	// Safety check: don't write the cache state unless we're in DeliverTx.
-	if mode != runTxModeDeliver {
-		return result
-	}
 
-	// only update state if all messages pass
-	if result.IsOK() {
+	// Attempt to execute all messages and only update state if all messages pass
+	// and we're in DeliverTx.
+	result, err = app.runMsgs(runMsgCtx, msgs, mode)
+	if err == nil && mode == runTxModeDeliver {
 		msCache.Write()
 	}
 
-	return result
+	return gInfo, result, err
 }
 
 // runMsgs iterates through all the messages and executes them.
-func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (result sdk.Result) {
+func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (*sdk.Result, error) {
 	msgLogs := make(sdk.ABCIMessageLogs, 0, len(msgs))
-
 	data := make([]byte, 0, len(msgs))
-	var (
-		code      sdk.CodeType
-		codespace sdk.CodespaceType
-	)
-
 	events := sdk.EmptyEvents()
 
-	// NOTE: GasWanted is determined by ante handler and GasUsed by the GasMeter.
+	// NOTE: GasWanted is determined by the AnteHandler and GasUsed by the GasMeter.
 	for i, msg := range msgs {
-		// match message route
+		// skip actual execution for (Re)CheckTx mode
+		if mode == runTxModeCheck || mode == runTxModeReCheck {
+			break
+		}
+
 		msgRoute := msg.Route()
 		handler := app.router.Route(msgRoute)
 		if handler == nil {
-			return sdk.ErrUnknownRequest("unrecognized message type: " + msgRoute).Result()
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unrecognized message route: %s", msgRoute)
 		}
 
-		var msgResult sdk.Result
-
-		// skip actual execution for CheckTx and ReCheckTx mode
-		if mode != runTxModeCheck && mode != runTxModeReCheck {
-			msgResult = handler(ctx, msg)
-		}
+		result, err := handler(ctx, msg)
 
 		// Each message result's Data must be length prefixed in order to separate
 		// each result.
-		data = append(data, msgResult.Data...)
-
-		msgEvents := msgResult.Events
+		data = append(data, result.Data...)
```

Comment:
I figured that we should expect exactly one of `result` or `err` is non-nil. That is suggested from:
`type Handler func(ctx Context, msg Msg) (*Result, error)`.
If `Result` must always be non-nil, please make it a struct rather than a pointer. At the least explicitly check if `result` is non-nil before reading any values from it.

My personal preference if just to return an error on a fail. If we have multiple messages, logs/data/events of the non-failure ones can be recorded, but I don't see why we expect any info from the failed message. This did bring up some confusion to other devs on slack as well.

That said, if you think it is better this way, that is fine. Just make sure to check or enforce the result is not nil

Reply:
Yes, great catch! We need to make `runMsgs` an exception -- it being a private method allows to developers not to have to worry about this.

Since we _always_ want to return a `Result` regardless of message execution success or failure, I agree we should change the signature to `sdk.Result, error` and always return a `Result`. This allows semantics to remain unchanged as `runTx` will not commit state upon `runMsgs` error.

Reply:
On further thought, this really makes semantics hard to understand. If I as the developer and enforcer of these semantics find it difficult, then surely it'll be difficult for upstream users.

Thus, I believe we should go with the simple and intuitive approach: always return `*Result` or `error`, but never both. This means that if any single message fails, we'll only retain data for that
failed message in the error -- losing any info on previous successful messages and I think this is OK. This is what you were trying to achieve the entire time @ethanfrey.